### PR TITLE
[FIX] Composer: select headers when composer is open

### PR DESCRIPTION
**Pull request by yunchen4**
## Description:

This PR fixes the usability problem that when the composer is open, the headers cannot be selected. 

Odoo task ID : [2835588](https://www.odoo.com/web#id=2835588&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/components/headers_overlay/headers_overlay.ts [hunk 1]
```diff
@@ -200,6 +200,10 @@ abstract class AbstractResizer extends Component<ResizerProps, SpreadsheetChildE
       this.startMovement(ev);
       return;
     }
+    if (this.env.model.getters.getEditionMode() === "editing") {
+      this.env.model.dispatch("UNFOCUS_SELECTION_INPUT");
+      this.env.model.dispatch("STOP_EDITION");
+    }
```
Comment (LucasLefevre) [inline src/components/headers_overlay/headers_overlay.ts:206]: as discussed, we should add a possibility on the selection processor to kick the active handler, whatever it is.
This part should not know any detail about who it is and how it works

File: src/components/grid/grid.ts [hunk 2]
```diff
@@ -471,8 +471,8 @@ export class Grid extends Component<Props, SpreadsheetChildEnv> {
     const lastZone = zones[zones.length - 1];
     let type: ContextMenuType = "CELL";
     if (!isInside(col, row, lastZone)) {
-      this.env.model.dispatch("UNFOCUS_SELECTION_INPUT");
-      this.env.model.dispatch("STOP_EDITION");
+      this.env.model.dispatch("STOP_EDITION"); // to keep the content
```
Comment (rrahir) [inline src/components/grid/grid.ts:474]: this should be handled by the `getBacktoDefault` function. the release callback set in the edition plugin ( https://github.com/odoo/o-spreadsheet/blob/master/src/plugins/ui_stateful/edition.ts#L418 ) should do the same thing that the `dispatch("stop_edition")` does. you can probably extract the logic in a single function.

File: src/selection_stream/event_stream.ts [hunk 3]
```diff
@@ -85,6 +85,14 @@ export class EventStream<Event> {
     this.mainSubscription = this.defaultSubscription;
   }
 
+  /**
+   * Release whichever subscription in charge and get back to the default subscription
+   */
+  getBackToDefault() {
+    this.mainSubscription?.callbacks.release?.();
```
Comment (rrahir) [inline src/selection_stream/event_stream.ts:92]: I wonder if we should *not* call the release when the mainsubscription is the same a the default one.

File: tests/components/composer.test.ts [hunk 4]
```diff
@@ -1206,6 +1207,26 @@ describe("composer", () => {
     expect(fixture.querySelectorAll(".o-grid div.o-composer")).toHaveLength(0);
   });
 
+  test("The composer should be closed before selecting headers", async () => {
+    composerEl = await typeInComposerGrid("=");
+    composerEl.dispatchEvent(new MouseEvent("click"));
```
Comment (rrahir) [inline tests/components/composer.test.ts:1212]: why is this necessary ? Don't you have the focus once you start typing?

File: tests/components/composer.test.ts [hunk 5]
```diff
@@ -1206,6 +1207,26 @@ describe("composer", () => {
...
+    expect(fixture.querySelectorAll(".o-grid-composer .o-composer")).toHaveLength(1);
+    await selectColumnByClicking(model, "C");
+    await nextTick();
```
Comment (rrahir) [inline tests/components/composer.test.ts:1215]: is the nextTick relevant here? I don't really see what it could change

File: src/plugins/ui/edition.ts [hunk 6]
```diff
@@ -401,7 +401,10 @@ export class EditionPlugin extends UIPlugin {
       { cell: { col: this.col, row: this.row }, zone },
       {
         handleEvent: this.handleEvent.bind(this),
-        release: () => (this.mode = "inactive"),
+        release: () => {
+          this.stopEdition();
+          this.mode = "inactive";
```
[the remark this answers is not in the document]
Comment (yunchen4) [reply src/plugins/ui/edition.ts:406]: > this is already taken care of in stopSelection :)

I didn't find any method called `stopSelection` in codebase :( which one did you refer to?

File: tests/components/composer.test.ts (hunk 4 above)
Comment (yunchen4) [reply tests/components/composer.test.ts]: Because I type in "=" I need to click the cell again to make it into editing mode (otherwise it is selecting mode). I've changed into "Hello" and this click event is deleted now

File: tests/components/composer.test.ts (hunk 5 above)
Comment (yunchen4) [reply tests/components/composer.test.ts]: Deleting this line will make the test fail. I've added `await nextTick()` in `selectColumnByClicking` helper and this line is deleted.

**Comment by rrahir** [comment]
@robodoo r+
